### DISCUSSION TITLE
What is the type of the dereferenced .items iterator?

### DISCUSSION THREAD

**Discussion by gantonopoulos** [nlohmann/json#2480] [Q&A] [answered]
I have the following json which I think maps to a std::map. So I try to iterato over it using std::for_each
but I cannot figure out the correct type of the dereferenced .items iterator, so as to put it as an input argument to my unary function (code below does not compile). What would that be?

`
json json_string = "{\"A\": {\n"
                              "        \"code\": \"0x00024182\",\n"                        
                              "        }\n"
                              "    },\n"
                              "    \"B\": {\n"
                              "        \"code\": \"0x00024A01\",\n"
                              "        }\n"
                              "    }}"_json;

    std::for_each(json_string.items().begin(),json_string.items().end(),[](std::pair<const json,json>& arg)
    {
        std::cout << arg.first.get<std::string>() << ":" << arg.second.get<std::string>() << "\n";
    });`

**Comment by nlohmann** [comment, marked as answer]
See https://github.com/nlohmann/json/discussions/2452

**Reply by gantonopoulos** [reply]
I really need read Cukic's book.

**Comment by gregmarr** [comment]
@gantonopoulos If you are using at least C++14, you can use `auto` arguments in your lambda.

    std::for_each(json_string.items().begin(),json_string.items().end(),[](auto const& arg)
    {
        std::cout << arg.first.get<std::string>() << ":" << arg.second.get<std::string>() << "\n";
    });`

**Reply by gantonopoulos** [reply]
I tried that as well but unfortunately we are using C++11.
